protocol/decoder/decoder.go

```diff
@@ -0,0 +1,271 @@
+	// Need some feedback on this
+	if variableName == "" {
+		return
+	}
+
+	indentationString := d.getIndentationString()
+
+	switch castedValue := value.(type) {
+	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+		d.logger.Debugf("%s%s (%d)", indentationString, variableName, castedValue)
+	case string:
+		d.logger.Debugf("%s%s (%s)", indentationString, variableName, castedValue)
+	case bool:
+		d.logger.Debugf("%s%s (%t)", indentationString, variableName, castedValue)
+	default:
+		d.logger.Debugf("%s%s (%v)", indentationString, variableName, castedValue)
+	}
+}
+
+func (d *Decoder) logTagBuffer() {
+	d.logger.Debugf("%s%s", d.getIndentationString(), "TAG_BUFFER")
+}
+
+// Primitive Types
+
+func (d *Decoder) ReadInt16(variableName string) (int16, error) {
+	if d.UnreadBytesCount() < 2 {
+		rem := d.UnreadBytesCount()
+		errorMessage := fmt.Sprintf("Expected int16 length to be 2 bytes, got %d bytes", rem)
+		return -1, errors.NewPacketDecodingError(errorMessage).AddContexts("INT16", variableName)
+	}
+
+	decodedInteger := int16(binary.BigEndian.Uint16(d.bytes[d.offset:]))
+	d.offset += 2
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+func (d *Decoder) ReadInt32(variableName string) (int32, error) {
+	if d.UnreadBytesCount() < 4 {
+		rem := d.UnreadBytesCount()
+		errorMessage := fmt.Sprintf("Expected int32 length to be 4 bytes, got %d bytes", rem)
+		err := errors.NewPacketDecodingError(errorMessage).AddContexts("INT32", variableName)
+		return -1, err
+	}
+
+	decodedInteger := int32(binary.BigEndian.Uint32(d.bytes[d.offset:]))
+	d.offset += 4
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+func (d *Decoder) ReadUnsignedVarint(variableName string) (uint64, error) {
+	decodedInteger, n := binary.Uvarint(d.bytes[d.offset:])
+
+	if n == 0 {
+		return 0, errors.NewPacketDecodingError("Unexpected end of data").AddContexts("UNSIGNED_VARINT", variableName)
+	}
+
+	if n < 0 {
+		errorMessage := fmt.Sprintf("Unexpected unsigned varint overflow after decoding %d bytes", -n)
+		return 0, errors.NewPacketDecodingError(errorMessage).AddContexts("UNSIGNED_VARINT", variableName)
+	}
+
+	d.offset += n
+	d.logDecodedValue(variableName, decodedInteger)
+	return decodedInteger, nil
+}
+
+// Composite Types : Uses Basic Types for decoding
+
+func (d *Decoder) ReadArrayLength(variableName string) (int, error) {
+	if d.UnreadBytesCount() < 4 {
+		rem := d.UnreadBytesCount()
+		errorMessage := fmt.Sprintf("Expected array length prefix to be 4 bytes, got %d bytes", rem)
+		return -1, errors.NewPacketDecodingError(errorMessage).AddContexts("ARRAY_LENGTH", variableName)
+	}
+	decodedInteger, err := d.ReadInt32("")
+
+	if err != nil {
+		if decodingErr, ok := err.(*errors.PacketDecodingError); ok {
+			return -1, decodingErr.AddContexts("ARRAY_LENGTH", variableName)
```

Comment:
This `Contexts` thing seems to account for a lot of code here. Would look a lot more simpler if this wasn't needed. I'd think a bit more about exactly what kind of error message we're trying to construct and see if what's needed could be stored as state on `Decoder`. 

Can be a separate PR might need some thinking re: approach